chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -13,705 +13,584 @@
-                &inputs,
-                &inputs_utxos,
-            )
-            .unwrap();
-
-        let result = constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap_err();
-        assert_eq!(
-            result,
-            IOPolicyError::AttemptToPrintMoneyOrViolateTimelockConstraints
-        );
+    /// Return accumulated coins that are left
+    pub fn consume(
+        self,
+        chain_config: &ChainConfig,
+        block_height: BlockHeight,
+    ) -> Result<Fee, IOPolicyError> {
+        let unconstrained_change = self
+            .unconstrained_value
+            .get(&CoinOrTokenId::Coin)
+            .cloned()
+            .unwrap_or(Amount::ZERO);
+
+        let maturity_distance = chain_config.staking_pool_spend_maturity_block_count(block_height);
+
+        let timelocked_change = self
+            .timelock_constrained
+            .into_iter()
+            .filter_map(|(lock, amount)| {
+                (lock.get() <= maturity_distance.to_int()).then_some(amount)
+            })
+            .sum::<Option<Amount>>()
+            .ok_or(IOPolicyError::CoinOrTokenOverflow(CoinOrTokenId::Coin))?;
 
-        // valid case
-        let outputs = vec![
-            TxOutput::LockThenTransfer(
-                OutputValue::Coin(Amount::from_atoms(staked_atoms - 10)),
-                Destination::AnyoneCanSpend,
-                OutputTimeLock::ForBlockCount(required_maturity_distance.to_int() as u64),
-            ),
-            TxOutput::LockThenTransfer(
-                OutputValue::Coin(Amount::from_atoms(10)),
-                Destination::AnyoneCanSpend,
-                OutputTimeLock::ForBlockCount(required_maturity_distance.to_int() as u64),
-            ),
-            TxOutput::Transfer(
-                OutputValue::Coin(less_than_staked_amount),
-                Destination::AnyoneCanSpend,
-            ),
-        ];
-
-        {
-            let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-            constraints_accumulator
-                .process_inputs(
-                    &chain_config,
-                    BlockHeight::new(1),
-                    pledge_getter,
-                    &inputs,
-                    &inputs_utxos,
-                )
-                .unwrap();
-
-            constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap();
+        let fee = (unconstrained_change + timelocked_change)
+            .ok_or(IOPolicyError::CoinOrTokenOverflow(CoinOrTokenId::Coin))?;
+
+        Ok(Fee(fee))
+    }
+}
+
+fn insert_or_increase<K: Ord>(
+    total_amounts: &mut BTreeMap<K, Amount>,
+    key: K,
+    amount: Amount,
+) -> Result<(), IOPolicyError> {
+    match total_amounts.entry(key) {
+        Entry::Occupied(mut entry) => {
+            let value = entry.get_mut();
+            *value = (*value + amount).ok_or(IOPolicyError::AmountOverflow)?;
+        }
+        Entry::Vacant(ventry) => {
+            ventry.insert(amount);
```

Comment:
Minor: I think the `Entry` API has methods for this pattern, something like this should do:
```rust
let value = total_amounts.entry(key).or_insert(Amount::ZERO);
*value = (*value + amount).ok_or(...)
```